File: chattools/src/main/java/aytackydln/chattools/telegram/dto/response/SendMessage.java
package aytackydln.chattools.telegram.dto.response;

import aytackydln.chattools.telegram.exception.TelegramException;
import aytackydln.chattools.telegram.dto.models.Keyboard;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.PropertyNamingStrategy;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import lombok.Data;
import lombok.ToString;

import java.util.Calendar;
import java.util.Date;

@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy.class)
@Data
@ToString(onlyExplicitlyIncluded = true)
public class SendMessage implements TelegramResponse {

    @ToString.Include
    private final long chatId;

    @ToString.Include
    private String text;
    private boolean disableNotification;
    private Keyboard replyMarkup;

    private boolean errorProcessed = false;

    public SendMessage(final long chatId, final String text) {
        this.chatId = chatId;
        this.text = text;
    }

    public SendMessage keyboard(final Keyboard keyboard) {
        replyMarkup = keyboard;
        return this;
    }

    public SendMessage silent() {
        disableNotification = true;
        return this;
    }

    public String getMethod() {
        return "sendMessage";
    }

    @JsonProperty
    public String getParseMode() {
        return "HTML";
    }

    @Override
    @JsonIgnore
    public boolean isLimited() {
        return true;
    }

    @Override
    public TelegramResponse onError(TelegramException e) {
        if (!errorProcessed) {
            errorProcessed = true;
            return this;
        }
        return null;
    }

    @Override
    public void preSend() {
        //dont make notification sound between 00:00-07:00
        final var calendar = Calendar.getInstance(); // creates a new calendar instance
        calendar.setTime(new Date());   // assigns calendar to given date
        final int hour = calendar.get(Calendar.HOUR_OF_DAY);
        if (hour < 7) {
            silent();
        }
    }
}

File: chattools/src/main/kotlin/aytackydln/chattools/telegram/dto/response/SendMessage.kt
package aytackydln.chattools.telegram.dto.response

import aytackydln.chattools.telegram.dto.models.Keyboard
import aytackydln.chattools.telegram.exception.TelegramException
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.PropertyNamingStrategies
import com.fasterxml.jackson.databind.PropertyNamingStrategy
import com.fasterxml.jackson.databind.annotation.JsonNaming
import lombok.Data
import lombok.ToString
import java.util.*

@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy::class)
@Data
@ToString(onlyExplicitlyIncluded = true)
data class SendMessage(
    @field:ToString.Include val chatId: Long,
    @field:ToString.Include val text: String
) :
    TelegramResponse {
    var disableNotification = false
    var replyMarkup: Keyboard? = null

    var errorProcessed = false

    fun keyboard(keyboard: Keyboard?): SendMessage {
        replyMarkup = keyboard
        return this
    }

    fun silent(): SendMessage {
        disableNotification = true
        return this
    }

    override val method: String
        get() = "sendMessage"

    @get:JsonProperty
    val parseMode: String
        get() = "HTML"

    @get:JsonIgnore
    override val isLimited: Boolean
        get() = true

    override fun onError(e: TelegramException): TelegramResponse? {
        if (!errorProcessed) {
            errorProcessed = true
            return this
        }
        return null
    }

    override fun preSend() {
        //dont make notification sound between 00:00-07:00
        val calendar = Calendar.getInstance() // creates a new calendar instance
        calendar.setTime(Date()) // assigns calendar to given date
        val hour = calendar.get(Calendar.HOUR_OF_DAY)
        if (hour < 7) {
            silent()
        }
    }
}
